salsa-2022-tests/tests/cycles.rs

```diff
@@ -0,0 +1,425 @@
+    cycle_b,
+    cycle_c,
+);
+
+trait Db: salsa::DbWithJar<Jar> {}
+
+#[salsa::db(Jar)]
+#[derive(Default)]
+struct Database {
+    storage: salsa::Storage<Self>,
+}
+
+impl salsa::Database for Database {
+    fn salsa_runtime(&self) -> &salsa::Runtime {
+        self.storage.runtime()
+    }
+}
+
+impl Db for Database {}
+
+impl RefUnwindSafe for Database {}
+
+#[salsa::input(jar = Jar)]
+struct MyInput {}
+
+#[salsa::tracked(jar = Jar)]
+fn memoized_a(db: &dyn Db, input: MyInput) {
+    memoized_b(db, input)
+}
+
+#[salsa::tracked(jar = Jar)]
+fn memoized_b(db: &dyn Db, input: MyInput) {
+    memoized_a(db, input)
+}
+
+#[salsa::tracked(jar = Jar)]
+fn volatile_a(db: &dyn Db, input: MyInput) {
+    db.runtime().report_untracked_read();
+    volatile_b(db, input)
+}
+
+#[salsa::tracked(jar = Jar)]
+fn volatile_b(db: &dyn Db, input: MyInput) {
+    db.runtime().report_untracked_read();
+    volatile_a(db, input)
+}
+
+/// The queries A, B, and C in `Database` can be configured
+/// to invoke one another in arbitrary ways using this
+/// enum.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum CycleQuery {
+    None,
+    A,
+    B,
+    C,
+    AthenC,
+}
+
+#[salsa::input(jar = Jar)]
+struct ABC {
+    a: CycleQuery,
+    b: CycleQuery,
+    c: CycleQuery,
+}
+
+impl CycleQuery {
+    fn invoke(self, db: &dyn Db, abc: ABC) -> Result<(), Error> {
+        match self {
+            CycleQuery::A => cycle_a(db, abc),
+            CycleQuery::B => cycle_b(db, abc),
+            CycleQuery::C => cycle_c(db, abc),
+            CycleQuery::AthenC => {
+                let _ = cycle_a(db, abc);
+                cycle_c(db, abc)
+            }
+            CycleQuery::None => Ok(()),
+        }
+    }
+}
+
+#[salsa::tracked(jar = Jar, recovery_fn=recover_a)]
+fn cycle_a(db: &dyn Db, abc: ABC) -> Result<(), Error> {
+    abc.a(db).invoke(db, abc)
+}
+
+fn recover_a(db: &dyn Db, cycle: &salsa::Cycle, abc: ABC) -> Result<(), Error> {
+    Err(Error {
+        cycle: cycle.all_participants(db),
+    })
+}
+
+#[salsa::tracked(jar = Jar, recovery_fn=recover_b)]
+fn cycle_b(db: &dyn Db, abc: ABC) -> Result<(), Error> {
+    abc.b(db).invoke(db, abc)
+}
+
+fn recover_b(db: &dyn Db, cycle: &salsa::Cycle, abc: ABC) -> Result<(), Error> {
+    Err(Error {
+        cycle: cycle.all_participants(db),
+    })
+}
+
+#[salsa::tracked(jar = Jar)]
+fn cycle_c(db: &dyn Db, abc: ABC) -> Result<(), Error> {
+    abc.c(db).invoke(db, abc)
+}
+
+#[track_caller]
+fn extract_cycle(f: impl FnOnce() + UnwindSafe) -> salsa::Cycle {
+    let v = std::panic::catch_unwind(f);
+    if let Err(d) = &v {
+        if let Some(cycle) = d.downcast_ref::<salsa::Cycle>() {
+            return cycle.clone();
+        }
+    }
+    panic!("unexpected value: {:?}", v)
+}
+
+#[test]
+fn cycle_memoized() {
+    let mut db = Database::default();
+    let input = MyInput::new(&mut db);
+    let cycle = extract_cycle(|| memoized_a(&db, input));
+    let expected = expect![[r#"
+        [
```

Comment:
We haven't implemented `debug` for [`DependencyKey`](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/components/salsa-2022/src/key.rs#L37)